File: frontend/src/lib/novelUnits.js
// Frontend mirror of app/services/domain/novel_units.py. The editor previews
// a generated key before anything is saved, so the fallback cannot live only
// on the server. src/config/novelUnitKinds.test.js pins the two together.

const KEY_PREFIX = {
  volume: "Vol",
  arc: "Arc",
  story: "Story",
  chapter: "Ch",
};

export const NOVEL_UNIT_KINDS_BY_TYPE = {
  "Light Novel": ["volume"],
  Novel: ["volume"],
  Web: ["arc"],
  Other: ["volume", "story", "chapter"],
};

export function unitDisplayKey(kind, position, unitKey) {
  if (unitKey && String(unitKey).trim()) return String(unitKey).trim();
  const prefix = KEY_PREFIX[kind] || "Unit";
  const pos = Number(position) || 0;
  return `${prefix} ${pos}`;
}

export function kindsForType(novelType) {
  return NOVEL_UNIT_KINDS_BY_TYPE[novelType] || ["volume"];
}

// Mirror of NOVEL_VOLUME_ONLY_TYPES in app/utils/constants.py, derived the
// same way: a type that may only hold volume rows counts volumes and nothing
// else. Its chapter and arc columns are not empty, they are meaningless, so
// derive_novel_progress() clears them server-side and nothing renders or
// edits them here.
export const NOVEL_VOLUME_ONLY_TYPES = Object.entries(NOVEL_UNIT_KINDS_BY_TYPE)
  .filter(([, kinds]) => kinds.length === 1 && kinds[0] === "volume")
  .map(([type]) => type);

/**
 * Whether this novel counts chapters (and, for Web, arcs) at all.
 *
 * An unrecognised or unset type counts them, matching the server: only the
 * two named volume-only types are cleared, so anything else keeps its pair.
 */
export function countsChapters(novel) {
  return !NOVEL_VOLUME_ONLY_TYPES.includes(novel?.type);
}

/**
 * Whether this novel shows a volume counter.
 *
 * Only Web says no. A web novel is read in chapters, and a volume number on
 * one is at best the print run it later got — so the columns are kept (a
 * type change brings them straight back) but nothing renders or edits them.
 * That asymmetry with countsChapters() is deliberate: chapters on a light
 * novel are meaningless and get cleared server-side, volumes on a web novel
 * are merely not the counter in use.
 */
export function countsVolumes(novel) {
  return novel?.type !== "Web";
}

function hasArcRows(novel) {
  return (novel?.units || []).some((u) => u.unit_kind === "arc");
}

const PROGRESS_DISPLAY_LABEL = {
  vol_original: "VOL JP/KR (Original Volumes)",
  vol_tw: "VOL TW (Taiwan Volumes)",
  ch: "CH (Chapters)",
  arc: "ARC (Arcs)",
  arc_ch: "ARC + CH (Arc and chapter)",
};

/**
 * The progress_display values this novel may be set to, as {value, label}.
 *
 * Built from what the entry actually has rather than from a flat list: a
 * volume counter only where volumes are counted, a chapter counter only
 * where chapters are, and the two arc counters only once arc rows exist —
 * picking "arc" on a novel with no arcs could render nothing but "?".
 *
 * The blank first option is the stored NULL, labelled with whatever
 * effectiveProgressDisplay() derives for this entry, so the admin can see
 * what "default" currently means before overriding it.
 */
export function progressDisplayOptions(novel) {
  const values = [];
  if (countsVolumes(novel)) values.push("vol_original", "vol_tw");
  if (countsChapters(novel)) values.push("ch");
  if (countsChapters(novel) && hasArcRows(novel)) values.push("arc", "arc_ch");

  const derived = PROGRESS_DISPLAY_LABEL[derivedProgressDisplay(novel)] || "";
  return [
    { value: "", label: `— Default (${derived}) —` },
    ...values.map((value) => ({ value, label: PROGRESS_DISPLAY_LABEL[value] })),
  ];
}

/**
 * Decision G: the type (plus, for Web, whether arc rows exist) determines
 * which progress pair a novel renders. `progress_display` narrows to a pure
 * override — when it is set, it wins outright and existing rows keep
 * rendering exactly as they did before this table existed (that is what
 * `withLegacyProgressDisplay` protects on the select side). When it is
 * unset, the mode is derived here so every consumer agrees:
 *
 * | novel.type                      | Effective mode        |
 * |----------------------------------|------------------------|
 * | Web, with arc rows                | "arc_ch" (two-stage)   |
 * | Web, no arc rows                  | "ch"                   |
 * | Light Novel / Novel                | "vol_original"          |
 * | Other, unit kind volume (or none)  | "vol_original"          |
 * | Other, unit kind story or chapter  | "ch"                   |
 *
 * Consumers: getNovelProgress (lib/formatters.js), MediaCard, and
 * NovelDashboardCard / NovelTrackerBlock all read this instead of the raw
 * `progress_display` column, so a web novel with arcs cannot show a volume
 * counter in one place and a two-stage row in another.
 */
function derivedProgressDisplay(novel) {
  const units = novel?.units || [];
  const type = novel?.type;

  if (type === "Web") {
    return hasArcRows(novel) ? "arc_ch" : "ch";
  }
  if (type === "Other") {
    return units.some((u) => u.unit_kind === "story" || u.unit_kind === "chapter")
      ? "ch"
      : "vol_original";
  }
  // Light Novel, Novel, and any unrecognised type default to volumes.
  return "vol_original";
}

export function effectiveProgressDisplay(novel) {
  const pd = novel?.progress_display;
  // The override only wins if the type still supports it. Pull and Fill write
  // this column, and a type change leaves an old value behind, so a Web novel
  // can hold "vol_tw" — honouring it would render a volume row the type does
  // not have. An unsupported value falls back to the derived mode; the select
  // still shows it (withLegacyProgressDisplay) so it is visible, not silently
  // swapped.
  if (pd && progressDisplayOptions(novel).some((o) => o.value === pd)) return pd;

  return derivedProgressDisplay(novel);
}

/**
 * One whole-arc step for the "arc" display: finish (or unfinish) an entire
 * arc and reset the in-arc cursor, so ch_fin stays exactly the sum of the
 * finished arcs. Clamped at both ends — unlike arcStep, which deliberately
 * runs past the last recorded arc for an ongoing novel, a whole-arc step has
 * no arc to move into.
 */
export function wholeArcStep(arcs, arcFin, chInArc, direction) {
  const total = (arcs || []).length;
  const fin = Math.max(0, Math.floor(Number(arcFin) || 0));
  const next = Math.min(total, Math.max(0, fin + direction));
  return { arc_fin: next, ch_fin_in_arc: 0 };
}

/**
 * One chapter step for a web novel with arcs, folded into the right arc.
 * Mirrors normalize_arc_progress: carrying stops at the last recorded arc,
 * because an ongoing novel is read into an arc nobody has entered yet.
 */
export function arcStep(arcs, arcFin, chInArc, direction) {
  const counts = (arcs || []).map((a) => Number(a.ch_count) || 0);
  let fin = Math.max(0, Math.floor(Number(arcFin) || 0));
  let ch = (Number(chInArc) || 0) + direction;

  while (ch < 0 && fin > 0) {
    fin -= 1;
    ch += counts[fin];
  }
  if (ch < 0) ch = 0;

  while (fin < counts.length) {
    const width = counts[fin];
    if (width <= 0 || ch < width) break;
    ch -= width;
    fin += 1;
  }

  return { arc_fin: fin, ch_fin_in_arc: ch };
}

File: app/services/domain/novel_units.py
"""
Novel unit arithmetic — rollover, derivation, and the display-key fallback.

Pure functions: no session, no queries. They are called from the router
(every write path) and from run_sync_novel (Calculate), so they must be
idempotent — running them twice on the same entry changes nothing.
"""

from app.utils.constants import NOVEL_UNIT_KEY_PREFIX, NOVEL_VOLUME_ONLY_TYPES


def _num(val) -> float:
    return float(val or 0)


def normalize_arc_progress(arc_counts, arc_fin, ch_fin_in_arc):
    """
    Fold an out-of-range chapter cursor into the right arc.

    arc_counts is every arc's ch_count in position order. arc_fin counts arcs
    that are *fully* finished, so the arc being read is at index arc_fin.

    Carrying up stops at the last recorded arc: an ongoing web novel is read
    into an arc nobody has entered yet, and clamping would discard that.
    An arc with an unknown (zero or null) count also stops the carry, because
    there is no width to subtract.
    """
    counts = [_num(c) for c in arc_counts]
    total = len(counts)
    fin = max(0, int(_num(arc_fin)))
    ch = _num(ch_fin_in_arc)

    # Borrow downward: a negative cursor means stepping back past the start
    # of the current arc, into the end of the previous one.
    while ch < 0 and fin > 0:
        fin -= 1
        ch += counts[fin]
    if ch < 0:
        ch = 0.0

    # Carry upward while the current arc has a known width and is full.
    while fin < total:
        width = counts[fin]
        if width <= 0 or ch < width:
            break
        ch -= width
        fin += 1

    return fin, ch


def clear_chapter_catalog(entry) -> None:
    """
    Blank the work's chapter and arc totals on a volume-only novel.

    A Light Novel or a Novel is counted in volumes, so these carry no meaning
    for it. Both are nullable, so both go to None.
    """
    entry.arc_total = None
    entry.ch_total = None


def clear_chapter_list(row) -> None:
    """The reader's half of the same clear: the counters go to 0, not None -
    they are a position, and position zero is where you are before you start."""
    row.arc_fin = 0
    row.ch_fin = 0
    row.ch_fin_in_arc = 0


def derive_novel_catalog(entry) -> None:
    """
    Recompute the work's arc_total and ch_total from its arc rows.

    Decision B: only arcs are authoritative. Volume rows are optional
    enrichment, so vol_total_original / vol_total_tw are never touched here.
    The type gates all of it - a volume-only type is cleared outright and
    never derives, even when arc rows are present, which a Pull can carry in.

    Pure: no session, no queries, and idempotent, because it is called from
    every write path and from Calculate.
    """
    if getattr(entry, "type", None) in NOVEL_VOLUME_ONLY_TYPES:
        clear_chapter_catalog(entry)
        return

    arcs = sorted(
        (u for u in (entry.units or []) if u.unit_kind == "arc"),
        key=lambda u: _num(u.position),
    )
    if not arcs:
        return

    counts = [_num(u.ch_count) for u in arcs]
    entry.arc_total = float(len(arcs))
    entry.ch_total = float(sum(counts))


def derive_novel_list(row, entry) -> None:
    """
    Recompute one reader's arc_fin, ch_fin_in_arc and ch_fin from the work's
    arc rows and their own cursor.

    The arithmetic is exactly what derive_novel_progress did; what changed is
    where the answer is written. Two people reading the same web novel fold
    their own cursors through the same arc widths and land in different
    places, which is the whole point.
    """
    if getattr(entry, "type", None) in NOVEL_VOLUME_ONLY_TYPES:
        clear_chapter_list(row)
        return

    arcs = sorted(
        (u for u in (entry.units or []) if u.unit_kind == "arc"),
        key=lambda u: _num(u.position),
    )
    if not arcs:
        # No arcs: the flat ch_fin the reader typed stands, and only the
        # in-arc cursor is meaningless. Same rule derive_novel_progress had.
        row.ch_fin_in_arc = 0
        return

    counts = [_num(u.ch_count) for u in arcs]
    fin, ch = normalize_arc_progress(counts, row.arc_fin, row.ch_fin_in_arc)

    row.arc_fin = float(fin)
    row.ch_fin_in_arc = float(ch)
    row.ch_fin = float(sum(counts[:fin]) + ch)


def unit_display_key(unit_kind, position, unit_key) -> str:
    """
    The label shown for a unit. An explicit unit_key always wins; otherwise
    generate one from kind and position ("Vol 1", "Arc 2"). Display-time only
    — the generated value is never stored.
    """
    if unit_key and str(unit_key).strip():
        return str(unit_key).strip()
    prefix = NOVEL_UNIT_KEY_PREFIX.get(unit_kind, "Unit")
    pos = _num(position)
    shown = int(pos) if pos == int(pos) else pos
    return f"{prefix} {shown}"
